Add Fastdotcom config flow

erwindouna opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adding a config flow the Fast.com integration. Also including unit testing for the config flow This is a split of https://github.com/home-assistant/core/pull/98602 which will be deleted later on. Acccording to Johansson's feedback I have splitted the Fast.com PR in three parts. This is part one only containing the config flow.
Removed some old behaviors, such as manual disabling, since this can be done in the new entity UI and put the default polling at one hour. If users require more updates I would suggest updating the documentation to tell users to utilize the update_entity service as they deem useful for their needs.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #98602
- This PR is related to issue: #98602 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28654

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

erwindouna committed:
Adding config flow and tests

erwindouna committed:
Removing update and adding to integrations.json

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,17 +36,17 @@
 
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+async def async_setup_platform(hass: HomeAssistant, config: ConfigType) -> bool:
```
This should only now import to the config flow. This still sets up the integration

gjohansson-ST replied:
Seems the import should be from the sensor so essentially this could be deleted and remaining should go into `setup_entry`

erwindouna replied:
Effectively meaning the entire async_setup_platform can be deleted in __init__.py?

gjohansson-ST replied:
Not yet as people can have a yaml config so it needs to be kept but can simply return true until the yaml config can be removed completely.

erwindouna replied:
Clear, thanks. Code has been changed! :)

erwindouna committed:
Updating hassfest

gjohansson-ST commented on homeassistant/components/fastdotcom/config_flow.py:
```diff
@@ -0,0 +1,35 @@
+            return self.async_abort(reason="single_instance_allowed")
+
+        if user_input is not None:
+            await self.async_set_unique_id(DOMAIN)
```
No need to set unique id

erwindouna replied:
Done.

gjohansson-ST replied:
Setting unique id still here?

erwindouna replied:
It should have been removed in this commit: https://github.com/home-assistant/core/pull/98686/commits/3f918fe031d71da1a8df135822b166c47a04fd75

erwindouna committed:
Removing comments

gjohansson-ST commented on homeassistant/components/fastdotcom/sensor.py:
```diff
@@ -8,21 +8,46 @@
     discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up the Fast.com sensor."""
+    async_create_issue(
```
We should have an import here and the issue should be created in config flow depending on outcome. See ADR-21

erwindouna replied:
Can you clarify on this a bit please? I see tons of examples where the create_issue for the deprecation is stored in sensor.py/light.py, etc. What did I miss here? :)

gjohansson-ST replied:
https://github.com/home-assistant/architecture/blob/master/adr/0021-YAML-integration-configuration-deprecation-policy.md
There are some PR's open so look at others now doing same thing as you do

erwindouna replied:
Got it - feedback has been processed.

gjohansson-ST reviewed (changes requested):
Some immediate feedback

erwindouna committed:
Removing unique ID

erwindouna committed:
Putting the setup_platform out of order

erwindouna committed:
Adding feedback on issues and importing

erwindouna commented:
Docs updated: https://github.com/home-assistant/home-assistant.io/pull/28637

gjohansson-ST commented:
Add it to the pr description in the right place. 
Thanks

erwindouna commented:
> Add it to the pr description in the right place. Thanks

Done and done.

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,38 +35,49 @@
 
-    hass.async_create_task(
-        async_load_platform(hass, Platform.SENSOR, DOMAIN, {}, config)
+    hass.data.setdefault(DOMAIN, {})
```
Before this saved `SpeedtestData(hass)` but no longer so does it actually work?

erwindouna replied:
Yes, it's fully operational, even with the split. Is there any evidence I should deliver?

gjohansson-ST replied:
No, found it so this line can be just removed

erwindouna replied:
I see. Are you sure you're checking the latest commit(s)? Whereas this shouldn't be in (anymore).

![image](https://github.com/home-assistant/core/assets/5011203/287b7301-6a39-48f1-a99d-a5f966fbf266)

gjohansson-ST replied:
`hass.data.setdefault(DOMAIN, {})` is still here which is not needed so it can be removed

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,38 +35,49 @@
 
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload Fast.com config entry."""
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```
See other integrations that clean `hass.data` before unloading

erwindouna replied:
Patched.

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,38 +35,49 @@
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
 class SpeedtestData:
```
In a follow-up PR I think it would be good to make this a proper `DataUpdaterCoordinator`

erwindouna replied:
Indeed, that is what my first PR had included. :)

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,38 +35,49 @@
 
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+async def async_setup_platform(hass: HomeAssistant, config: ConfigType) -> bool:
```
We need to move the import here as the user will never come to the platform.

erwindouna replied:
Your intital feedback was:

> Not yet as people can have a yaml config so it needs to be kept but can simply return true until the yaml config can be removed completely.

Should the create_task be moved here, as well as a redirect to the `async_step_import`?

gjohansson-ST replied:
According to the documentation the integration loads from init so that's were we should start the import so all of it should be moved there.

gjohansson-ST commented on homeassistant/components/fastdotcom/sensor.py:
```diff
@@ -8,21 +8,37 @@
     discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Old legacy setup. Now initialize the import."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            FASTDOTCOM_DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data={},
+        )
+    )
```
Move to `__init__.py` and remove `async_setup_platform()`

gjohansson-ST commented on homeassistant/components/fastdotcom/strings.json:
```diff
@@ -1,8 +1,13 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Do you want to start the setup? The initial setup will take about 30-40 seconds."
```
Seems strange? Setting up the config entry is pretty instant. Or you mean first data refresh which is not relevant here.

erwindouna replied:
Yes, I thought it might be helpful to notify the user that the initial data refresh takes longer, since it will actually perform it twice. Each call takes around 15-16 seconds. So to hint out that HA isn't crashing, just a headsup the initialization process takes a bit more time.

erwindouna committed:
Merge branch 'home-assistant:dev' into fastdotcom-configflow-only

erwindouna committed:
Removing uniqueID (again)

erwindouna committed:
Adjusting unload and typo

erwindouna committed:
Updating manifest properly

erwindouna committed:
Minor patching

erwindouna committed:
Removing hass.data.setdefault(DOMAIN, {})

erwindouna committed:
Moving load_platform to __init__.py

gjohansson-ST commented on homeassistant/components/fastdotcom/config_flow.py:
```diff
@@ -0,0 +1,51 @@
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
```
We need a separate issue if import fails because of this

erwindouna replied:
Can you help me out a bit what you mean by this?

frenck replied:
For import, this is correct behavior.

gjohansson-ST commented on homeassistant/components/fastdotcom/config_flow.py:
```diff
@@ -0,0 +1,51 @@
+            return self.async_abort(reason="single_instance_allowed")
+
+        if user_input is not None:
+            self._abort_if_unique_id_configured()
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/fastdotcom/strings.json:
```diff
@@ -1,8 +1,13 @@
+      }
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
```
```suggestion
```

erwindouna committed:
Update homeassistant/components/fastdotcom/config_flow.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

erwindouna committed:
Update homeassistant/components/fastdotcom/strings.json

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

frenck commented:
@erwindouna 👋  It seems like you have been addressing review comments and the CI seems to pass. Are you still working on things? Or is it ready for review? In the latter case, please don't forget to mark it "ready for review" so your PR gets out of the draft state.

../Frenck

erwindouna commented:
> @erwindouna 👋 It seems like you have been addressing review comments and the CI seems to pass. Are you still working on things? Or is it ready for review? In the latter case, please don't forget to mark it "ready for review" so your PR gets out of the draft state.
> 
> ../Frenck

I was awaiting the feedback from @gjohansson-ST. The CI indeed passes and it should be finished if you'd ask me. Maybe it slipped off the radar from Johansson - I'll mark it ready for review again. :)

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,38 +35,59 @@
 
-    if not conf[CONF_MANUAL]:
-        async_track_time_interval(hass, data.update, conf[CONF_SCAN_INTERVAL])
+    async_track_time_interval(hass, data.update, timedelta(hours=DEFAULT_INTERVAL))
```
We should stop this if the entry get's unloaded

erwindouna replied:
Fixed in the latest commit.

gjohansson-ST commented on homeassistant/components/fastdotcom/__init__.py:
```diff
@@ -40,38 +35,59 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload Fast.com config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
```
```suggestion
        hass.data.pop(DOMAIN)
```

gjohansson-ST commented on homeassistant/components/fastdotcom/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+            self.hass,
+            HOMEASSISTANT_DOMAIN,
+            f"deprecated_yaml_{DOMAIN}",
+            breaks_in_ha_version="2024.2.0",
```
```suggestion
            breaks_in_ha_version="2024.5.0",
```

gjohansson-ST reviewed (changes requested):
Some minor comments left

erwindouna committed:
Update homeassistant/components/fastdotcom/__init__.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

erwindouna committed:
Update homeassistant/components/fastdotcom/config_flow.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

erwindouna committed:
Adding an unload function for the timer

gjohansson-ST commented on homeassistant/components/fastdotcom/sensor.py:
```diff
@@ -8,21 +8,20 @@
 
 
-async def async_setup_platform(
+async def async_setup_entry(
```
I think if we remove the `setup_platform()` now we will get a separate repair issue about the YAML.
Can you test this if that's the case and restore an empty `setup_platform()` in that case?

erwindouna replied:
My test didn't showed any weird behavior, yet I just leave in an empty function just to be 100% sure. Hence, you hinted on this, so I suppose there are use cases it might conflicts randonmly. An empty function might prevent this.

erwindouna replied:
Pylint went haywire on an empty function. I inspired myself of the [NextBus](https://github.com/home-assistant/core/blob/2531b0bc09dba8d85243e8b7a4942840ef7b64d5/homeassistant/components/nextbus/sensor.py#L41), to include the create_issue here.
I recommend removing the create issue from the config flow.

gjohansson-ST reviewed (changes requested):
Just a minor question left to check

erwindouna committed:
Adding issue on setup platform in sensor

erwindouna commented:
@gjohansson-ST can I bother you checking this?

gjohansson-ST commented on homeassistant/components/fastdotcom/sensor.py:
```diff
@@ -8,14 +8,25 @@
+    async_add_entities([SpeedtestSensor(hass.data[FASTDOTCOM_DOMAIN])])
 
 
 async def async_setup_platform(
```
The complete `async_setup_platform()` function shall be removed. Don't leave it as empty either, just delete it all.
It shall not be used as this integration was only loaded from `__init__.py`

gjohansson-ST commented on homeassistant/components/fastdotcom/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+            self.hass,
+            HOMEASSISTANT_DOMAIN,
+            f"deprecated_yaml_{DOMAIN}",
+            breaks_in_ha_version="2024.5.0",
```
```suggestion
            breaks_in_ha_version="2024.6.0",
```

gjohansson-ST reviewed (changes requested):
Looks finished besides these two minor things

erwindouna committed:
Update homeassistant/components/fastdotcom/config_flow.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

erwindouna committed:
Removing platform

erwindouna committed:
Fixing strings.json

erwindouna committed:
Fine-tuning

gjohansson-ST commented on homeassistant/components/fastdotcom/sensor.py:
```diff
@@ -52,10 +51,8 @@ async def async_added_to_hass(self) -> None:
-            return
-        self._attr_native_value = state.state
+        # Initial run to get data
+        self.update()
```
This would block it for completion as it takes a long time.
I think we should go back to restoring the previous value and let it update on it's regular cycle

gjohansson-ST reviewed (changes requested):
Not sure if I missed this one before or you changed that just now?

erwindouna committed:
Putting back last_state

gjohansson-ST commented:
Test failure unrelated

gjohansson-ST reviewed (approved):
Thanks @erwindouna 👍

gjohansson-ST commented on homeassistant/components/fastdotcom/sensor.py:
```diff
@@ -8,29 +8,28 @@
 
 
 # pylint: disable-next=hass-invalid-inheritance # needs fixing
 class SpeedtestSensor(RestoreEntity, SensorEntity):
```
Would be good in a follow-up PR to change `RestoreEntity` to `RestoreSensor`

erwindouna replied:
Good point!
